packages/runtime/test-runtime-utils/src/mocks.ts

```diff
@@ -124,47 +165,124 @@ export class MockContainerRuntime {
 		// Set FluidDataStoreRuntime's deltaManager to ours so that they are in sync.
 		this.dataStoreRuntime.deltaManager = this.deltaManager;
 		this.dataStoreRuntime.quorum = factory.quorum;
+		this.dataStoreRuntime.containerRuntime = this;
 		// FluidDataStoreRuntime already creates a clientId, reuse that so they are in sync.
 		this.clientId = this.dataStoreRuntime.clientId ?? uuid();
 		factory.quorum.addMember(this.clientId, {});
+		this.runtimeOptions = makeContainerRuntimeOptions(mockContainerRuntimeOptions);
+		assert(
+			this.runtimeOptions.flushMode !== FlushMode.Immediate ||
+				!this.runtimeOptions.enableGroupedBatching,
+			"Grouped batching is not compatible with FlushMode.Immediate",
+		);
 	}
 
 	public createDeltaConnection(): MockDeltaConnection {
-		const deltaConnection = new MockDeltaConnection(
-			(messageContent: any, localOpMetadata: unknown) =>
-				this.submit(messageContent, localOpMetadata),
-			() => this.dirty(),
+		assert(
+			this.dataStoreRuntime.createDeltaConnection !== undefined,
+			"Unsupported datastore runtime version",
 		);
+		const deltaConnection = this.dataStoreRuntime.createDeltaConnection();
 		this.deltaConnections.push(deltaConnection);
 		return deltaConnection;
 	}
 
 	public submit(messageContent: any, localOpMetadata: unknown): number {
-		const clientSequenceNumber = this.clientSequenceNumber++;
-		const msg: Partial<ISequencedDocumentMessage> = {
-			clientId: this.clientId,
-			clientSequenceNumber,
-			contents: messageContent,
-			referenceSequenceNumber: this.deltaManager.lastSequenceNumber,
-			type: MessageType.Operation,
-		};
-		this.factory.pushMessage(msg);
+		const clientSequenceNumber = this.clientSequenceNumber;
+		switch (this.runtimeOptions.flushMode) {
+			case FlushMode.Immediate: {
+				this.factory.pushMessage({
+					clientId: this.clientId,
+					clientSequenceNumber,
+					contents: messageContent,
+					referenceSequenceNumber: this.referenceSequenceNumber,
+					type: MessageType.Operation,
+				});
+				this.addPendingMessage(messageContent, localOpMetadata, clientSequenceNumber);
+				this.clientSequenceNumber++;
+				break;
+			}
 
-		this.addPendingMessage(messageContent, localOpMetadata, clientSequenceNumber);
+			case FlushMode.TurnBased: {
+				this.outbox.push({
+					content: messageContent,
+					localOpMetadata,
+				});
+				break;
+			}
+
+			default:
+				throw new Error(`Unsupported FlushMode ${this.runtimeOptions.flushMode}`);
+		}
 
 		return clientSequenceNumber;
 	}
 
 	public dirty(): void {}
 
+	/**
+	 * If flush mode is set to FlushMode.TurnBased, it will send all messages queued since the last time
+	 * this method was called. Otherwise, calling the method does nothing.
+	 */
+	public flush?() {
```

Comment:
The reason for these methods to be optional is because of back-compat. Making them required will break backwards compatibility and we'd have to be explicit about it in package.json, which is not something we like to do in `main`. After merging, a corresponding PR will be made to `next` which will have these methods appropriately declared and also, we'll get rid of all the other asserts in the code related to this.

Reply:
I understand that when it comes to interface definition. But I'm not sure what "?" helps with here. It basically says that MockContainerRuntime  and derived classes could overwrite this property to be empty. Is this what we want?

Reply:
Not sure I follow, @vladsud. Which interface? This class is exported as-is in the public API. Clients are taking a dependency is on the class, take a look at index.ts. If this method is not optional, we'll technically break forward-compat and we'd need the 

```
"ClassDeclaration_MockContainerRuntimeForReconnection": {"forwardCompat": false}
"ClassDeclaration_MockContainerRuntime": {"forwardCompat": false}
```
in `package.json`. Not a big deal but also, it's something we try not to do in main if we can help it.

Do you have an alternative in mind?